sdk/src/utils/graphs/adjacency-pool-graph.ts

```diff
@@ -0,0 +1,195 @@
+    );
+
+    const results = searchTokenPairsAddrs.map(([startMint, endMint]) => {
+      const searchRouteId = PoolGraphUtils.getSearchRouteId(startMint, endMint);
+      const internalRouteId = getInternalRouteId(startMint, endMint);
+      const routesForSearchPair = walkMap[internalRouteId];
+      const paths = routesForSearchPair ? routesForSearchPair.map<Route>((route) => {
+        return {
+          startTokenMint: startMint,
+          endTokenMint: endMint,
+          hops: getHopsFromRoute(internalRouteId, searchRouteId, route),
+        };
+      }) : [];
+
+
+      return [searchRouteId, paths] as const;
+    });
+
+    return results;
+  }
+}
+
+
+
+function getHopsFromRoute(internalRouteId: string, searchRouteId: string, route: string[]): Hop[] {
+  const [intStartA] = PoolGraphUtils.deconstructRouteId(internalRouteId);
+  const [searchStartA] = PoolGraphUtils.deconstructRouteId(searchRouteId);
+  const shouldReverseRoute = searchStartA !== intStartA;
+  const finalRoutes = shouldReverseRoute ? route.reverse() : route;
+
+  return finalRoutes.map(hopStr => { return { poolAddress: new PublicKey(hopStr) } })
+}
+
+
+type AdjacencyPoolGraphMap = Record<string, Array<PoolGraphEdge>>;
+
+type PoolGraphEdge = {
+  address: string;
+  otherToken: string;
+};
+
+// A record of route-id (tokenA-tokenB) to a list of edges
+type PoolWalks = Record<string, string[][]>;
+
+function buildPoolGraph(pools: PoolTokenPair[]) {
+  const poolGraphSet = pools.reduce((poolGraph: Record<string, PoolGraphEdge[]>, pool) => {
+    const { address, tokenMintA, tokenMintB } = pool;
+    const [addr, mintA, mintB] = AddressUtil.toStrings([address, tokenMintA, tokenMintB]);
+
+    if (poolGraph[mintA] === undefined) {
+      poolGraph[mintA] = [];
+    }
+
+    if (poolGraph[mintB] === undefined) {
+      poolGraph[mintB] = [];
+    }
+
+    const existingAddressesForA = poolGraph[mintA].map((p) => p.address);
+    const existingAddressesForB = poolGraph[mintB].map((p) => p.address);
+
+
+    if (!existingAddressesForA.includes(addr)) {
+      poolGraph[mintA].push({ address: addr, otherToken: mintB });
+    }
+
+    if (!existingAddressesForB.includes(addr)) {
+      poolGraph[mintB].push({ address: addr, otherToken: mintA });
+    }
+
+    return poolGraph;
+  }, {});
+
+  return Object.fromEntries(
+    Object.entries(poolGraphSet).map(([mint, otherMints]) => [mint, Array.from(otherMints)])
+  );
+}
+
+// This is currently hardcoded to find walks of max length 2, generalizing to longer walks
+// may mean that a adjacency matrix might have better performance
+function findWalks(
+  tokenPairs: (readonly [string, string])[],
+  poolGraph: AdjacencyPoolGraphMap,
+  intermediateTokens?: string[]
+) {
+  const walks: PoolWalks = {};
+
+  tokenPairs.forEach(([tokenMintFrom, tokenMintTo]) => {
+    let routes = [];
+
+    const internalRouteId = getInternalRouteId(tokenMintFrom, tokenMintTo)
+    const poolsForTokenFrom = poolGraph[tokenMintFrom] || [];
```

Comment:
If we adjust the from/to from deconstructRouteId at this point, there would be no need to reverse at the end.

Reply:
nice!